url.go

```diff
@@ -473,14 +525,37 @@ func parseIdentifierURL(url *gourl.URL) (*URL, error) {
 	}
 
 	parts := strings.Split(strings.Trim(path, "/"), "/")
-	if len(parts) != 1 {
-		return nil, errors.Errorf(`charm or bundle URL %q malformed, expected "<name>"`, url)
+	if len(parts) == 0 || len(parts) > 3 {
+		return nil, errors.Errorf(`charm or bundle URL %q malformed`, url)
 	}
 
-	r.Name, r.Revision = extractRevision(parts[0])
+	var nameRev string
+	if num := len(parts); num == 3 {
+		r.Architecture, r.Series, nameRev = parts[0], parts[1], parts[2]
+	} else if num == 2 {
+		r.Architecture, nameRev = parts[0], parts[1]
+	} else {
+		nameRev = parts[0]
+	}
+
+	// Mandatory
+	r.Name, r.Revision = extractRevision(nameRev)
 	if err := ValidateName(r.Name); err != nil {
-		return nil, errors.Annotatef(err, "cannot parse URL %q", url)
+		return nil, errors.Annotatef(err, "cannot parse name and/or revision in URL %q", url)
```

Comment:
I always annotate with what we were doing, say `parsing name and/or revision in URL`. The fact that we couldn't do it is implied by the error.

Reply:
Agreed, but this is the convention in this package, so I'll keep it for now.